snap7/util.py

```diff
@@ -563,118 +586,123 @@ def get_offset(self, byte_index: Union[str, int]) -> int:
 
-        elif _type == 'REAL':
+        elif type_ == 'REAL':
             return get_real(bytearray_, byte_index)
 
-        elif _type == 'DWORD':
+        elif type_ == 'DWORD':
             return get_dword(bytearray_, byte_index)
 
-        elif _type == 'DINT':
+        elif type_ == 'DINT':
             return get_dint(bytearray_, byte_index)
 
-        elif _type == 'INT':
+        elif type_ == 'INT':
             return get_int(bytearray_, byte_index)
 
-        elif _type == 'WORD':
+        elif type_ == 'WORD':
             return get_word(bytearray_, byte_index)
 
-        elif _type == 'S5TIME':
+        elif type_ == 'S5TIME':
             data_s5time = get_s5time(bytearray_, byte_index)
             return data_s5time
 
-        elif _type == 'DATE_AND_TIME':
+        elif type_ == 'DATE_AND_TIME':
             data_dt = get_dt(bytearray_, byte_index)
             return data_dt
 
-        elif _type == 'USINT':
+        elif type_ == 'USINT':
             return get_usint(bytearray_, byte_index)
 
-        elif _type == 'SINT':
+        elif type_ == 'SINT':
             return get_sint(bytearray_, byte_index)
 
         # add these three not implemented data typ to avoid
         # 'Unable to get repr for class<snap7.util.DB_ROW>' error
-        elif _type == 'TIME':
+        elif type_ == 'TIME':
             return 'read TIME not implemented'
 
-        elif _type == 'DATE':
+        elif type_ == 'DATE':
             return 'read DATE not implemented'
 
-        elif _type == 'TIME_OF_DAY':
+        elif type_ == 'TIME_OF_DAY':
             return 'read TIME_OF_DAY not implemented'
 
         raise ValueError
 
-    def set_value(self, byte_index, _type, value):
+    def set_value(self, byte_index: Union[str, int], type: str, value: Union[bool, str, int, float]) -> Union[bytearray, None]:
         bytearray_ = self.get_bytearray()
 
-        if _type == 'BOOL':
+        if type == 'BOOL' and isinstance(value, bool):
             """
             mypy conform style:
             if isinstance(byte_index, str):
                 byte_index, bool_index = byte_index.split('.')
                 return set_bool(bytearray_, self.get_offset(byte_index),
                                 int(bool_index), value)
             """
-            byte_index, bool_index = byte_index.split(".")
+            byte_index, bool_index = str(byte_index).split(".")
             return set_bool(bytearray_, self.get_offset(byte_index),
                             int(bool_index), value)
 
         byte_index = self.get_offset(byte_index)
 
-        if _type.startswith('STRING'):
-            max_size = re.search(r'\d+', _type).group(0)
-            max_size = int(max_size)
-            return set_string(bytearray_, byte_index, value, max_size)
+        if type.startswith('STRING') and isinstance(value, str):
+            max_size = re.search(r'\d+', type)
+            if max_size is None:
```

Comment:
Not sure what the better handling would be here.
I would say, if it is just STRING, we use the default of 254. (Maybe for STRING[] as well with a warning?).
If it is STRING[123] we take the number.
If it is STRING[GiBbErIsCh] or the number is not in range between 1 and 254, raise an error.

For longer strings, an extra datatype exist, which we don't support atm.